sodetlib/operations/bias_dets.py

```diff
@@ -201,3 +201,408 @@ def bias_to_rfrac(S, cfg, rfrac=0.5, bias_groups=None, iva=None,
+    if bias_groups is None:
+        bias_groups = cfg.dev.exp['active_bgs']
+    bias_groups = np.atleast_1d(bias_groups)
+
+    ## take the initial biasstep
+    S.log("taking the first biasstep")
+    S.log(f"Initial dc biases {S.get_tes_bias_bipolar_array()}")
+    bsa_0 = bias_steps.take_bias_steps(S, cfg)
+    if make_plots:
+        fig, ax = bias_steps.plot_Rfrac(bsa_0)
+        fname = sdl.make_filename(S, 'intial_Rfrac.png', plot=True)
+        fig.savefig(fname)
+        S.pub.register_file(fname, 'biasstep_rfrac', format='png', plot=True)
+        if not show_plots:
+            plt.close(fig)
+            
+            
+    ## load IV analysis result so can use dynamic step size
+    iva = iv.IVAnalysis.load(bsa_0.meta['iv_file'])
+                
+    repeat_biasstep = False
+    ## check for script failue
+    for bl in bias_groups:
+        mask_bg = np.where(bsa_0.bgmap==bl)[0]
+        per_bl_R0 = bsa_0.R0[mask_bg]
+        if np.isnan(np.nanmedian(per_bl_R0)):
+            ## the yield is very low, could it be the biasstep analysis script failed? repeat again
+            repeat_biasstep = True
+            S.log("the biasstep script might have failed, going to repeat the measurement")
+
+    if repeat_biasstep:
+        S.log("retaking biasstep due to failure")
+        bsa_0 = bias_steps.take_bias_steps(S, cfg)
+        repeat_biasstep = False
+    
+    percentage_rn_0 = bsa_0.R0/bsa_0.R_n_IV
+    ## add in checks if detectors are SC  
+    bg_overbias_needed = []
+    overbias_needed = False
+    for bl in bias_groups:
+        mask_bg = np.where(bsa_0.bgmap==bl)[0]
+        per_bl_percentage_rn_0 = percentage_rn_0[mask_bg]
+        ## mask for SC
+        mask_sc = np.where((per_bl_percentage_rn_0 < 0.1) & (per_bl_percentage_rn_0 > -0.1))
+        ## if more than 10 percent of the detectors are SC
+        if len(mask_sc[0]) > 0.1*len(mask_bg):
+            bg_overbias_needed.append(bl)
+            overbias_needed = True
+
+    ## add in check if detectors are normal, hence we should increase the drop step
+    bg_detectors_normal = []
+    drop_from_normal = False
+    for bl in bias_groups:
+        mask_bg = np.where(bsa_0.bgmap==bl)[0]
+        per_bl_percentage_rn_0 = percentage_rn_0[mask_bg]
+        ## mask for normal
+        mask_normal = np.where((per_bl_percentage_rn_0 > 0.9) | (per_bl_percentage_rn_0 < -0.9))
+        ## if more than half of the detectors are normal
+        if len(mask_normal[0]) > 0.5*len(mask_bg):
+            bg_detectors_normal.append(bl)   
+            drop_from_normal = True
+
+
+    if overbias_needed:
+        S.log(f"some BL: {bg_overbias_needed} are stuck in SC, going to over bias them, this takes about 5mins")
+        previous_dc_biases = S.get_tes_bias_bipolar_array()
+        
+        sdl.overbias_dets(S, cfg, bias_groups=bg_overbias_needed)
+        ##force wait time 5 mins for UFM to reach equalibrium 
+        for sleep_index in range(5):
+            time.sleep(60)
```

Comment:
I think the sleep-time after overbiasing should be a configurable parameter, ideally settable in the dev-cfg

Reply:
yeah good idea.